datasetinsights/io/downloader/unity_simulation.py

```diff
@@ -0,0 +1,120 @@
+import logging
+import os
+import re
+
+from datasetinsights.io.downloader.base import DatasetDownloader
+from datasetinsights.io.usim import Downloader, download_manifest
+
+logger = logging.getLogger(__name__)
+
+
+class UnitySimulationDownloader(DatasetDownloader, protocol="usim://"):
+    """ This class is used to download data from Unity Simulation
+
+        For more on Unity Simulation please see these
+        `docs <https://github.com/Unity-Technologies/Unity-Simulation-Docs>`
+
+        Args:
+        access_token: "Access token to be used to authenticate to
+         unity simulation for downloading the dataset"
+
+    """
+
+    SOURCE_URI_PATTERN = (
+        r"usim://([^@]*)?@?"
+        r"([a-fA-F0-9]{8}-"
+        r"[a-fA-F0-9]{4}-"
+        r"[a-fA-F0-9]{4}-"
+        r"[a-fA-F0-9]{4}-"
+        r"[a-fA-F0-9]{12})"
+        r"/(\w+)"
+    )
+
+    def __init__(self, access_token=None, **kwargs):
+        """
+
+        Args:
+            access_token: "Access token to be used to authenticate to
+            unity simulation for downloading the dataset"
+        """
+        self.access_token = access_token
+        self.run_execution_id = None
+        self.project_id = None
+
+    def download(self, source_uri, output, include_binary=False, **kwargs):
+        """
+
+        Args:
+            source_uri: This is the downloader-uri that indicates where on
+                unity simulation the dataset should be downloaded from.
+
+                The expected source-uri follows these patterns
+
+                usim://access-token@project-id/run-execution-id
+                or
+                usim://project-id/run-execution-id
+
+
+            output: This is the path to the directory
+                where the download will store the dataset.
+
+            include_binary: Whether to download binary files
+                such as images or LIDAR point
+                clouds. This flag applies to Datasets where metadata
+                (e.g. annotation json, dataset catalog, ...)
+                 can be separated from binary files.
+
+        """
+        self.parse_source_uri(source_uri)
+        manifest_file = os.path.join(output, f"{self.run_execution_id}.csv")
+        if self.access_token:
```

Comment:
If `access_token` is not found or if it's not valid, Maybe it's better to raise exception?

Reply:
This is not part of this review. its directly taken from the previous implementation
   logger.info(
                f"No auth token is provided. Assuming you already have "
                f"a manifest file located in {manifest_file}"
            )

Reply:
I don't agree this is not part of the review. This seems to be an opportunity to fix "cache", or think about how we want to handle such cases. I am simply pointing out in the desired behavior, what should we do "If access_token is not found or if it's not valid". What's the desired behavior? What information is reverent and you want user to know? What should pass/fail?

Reply:
I'm trying to point out that this code was already checked into master i just moved it. So it should have been review before

Reply:
This was the original code in the usim_download.py script 
```

def run(args):
    if args.verbose:
        root_logger = logging.getLogger()
        root_logger.setLevel(logging.DEBUG)

    logger.info("Run command with args: %s\n", args)
    data_root = args.data_root

    auth_token = args.auth_token
    use_cache = not args.no_cache
    run_execution_id = args.run_execution_id

    manifest_file = os.path.join(
        data_root, const.SYNTHETIC_SUBFOLDER, f"{run_execution_id}.csv"
    )
    if auth_token:
        manifest_file = download_manifest(
            run_execution_id,
            manifest_file,
            auth_token,
            project_id=args.project_id,
            use_cache=use_cache,
        )
    else:
        logger.info(
            f"No auth token is provided. Assuming you already have "
            f"a manifest file located in {manifest_file}"
        )

    subfolder = Path(manifest_file).stem
    root = os.path.join(data_root, const.SYNTHETIC_SUBFOLDER, subfolder)

    dl_worker = Downloader(manifest_file, root, use_cache=use_cache)
    dl_worker.download_references()
    dl_worker.download_metrics()
    dl_worker.download_captures()
    if args.include_binary:
        dl_worker.download_binary_files()
```